src/wallet/sqlite.cpp

```diff
@@ -80,6 +81,72 @@ void SQLiteDatabase::Cleanup() noexcept
 
 void SQLiteDatabase::Open(const char* mode)
 {
+    int flags = SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE;
+    if (m_mock) {
+        flags |= SQLITE_OPEN_MEMORY; // In memory database for mock db
+    }
+
+    if (m_db == nullptr) {
+        TryCreateDirectories(m_dir_path);
+        int ret = sqlite3_open_v2(m_file_path.c_str(), &m_db, flags, nullptr);
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to open database: %s\n", sqlite3_errstr(ret)));
+        }
+    }
+
+    if (sqlite3_db_readonly(m_db, "main") != 0) {
+        throw std::runtime_error("SQLiteDatabase: Database opened in readonly mode but read-write permissions are needed");
+    }
+
+    // Acquire an exclusive lock on the database
+    // First change the locking mode to exclusive
+    int ret = sqlite3_exec(m_db, "PRAGMA locking_mode = exclusive", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error(strprintf("SQLiteDatabase: Unable to change database locking mode to exclusive: %s\n", sqlite3_errstr(ret)));
+    }
+    // Now begin a transaction to acquire the exclusive lock. This lock won't be released until we close because of the exclusive locking mode.
+    ret = sqlite3_exec(m_db, "BEGIN EXCLUSIVE TRANSACTION", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error("SQLiteDatabase: Unable to obtain an exclusive lock on the database, is it being used by another bitcoind?\n");
+    }
+    ret = sqlite3_exec(m_db, "COMMIT", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error(strprintf("SQLiteDatabase: Unable to end exclusive lock transaction: %s\n", sqlite3_errstr(ret)));
+    }
```

Comment:
f4fd2218d705ed08f8554f9cd1946c087335d2f2

According to SQLite [docs](https://sqlite.org/lockingv3.html#transaction_control) the actual moment of `EXCLUSIVE` lock acquiring is a bit vague:
> Note that the BEGIN command does not acquire any locks on the database. ... No EXCLUSIVE lock is acquired until either the memory cache fills up and must be spilled to disk or until the transaction commits.

> The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on. Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.

It's just a note for other reviewers.
>